### PR TITLE
[jit] C++ API Wrappers Proposal

### PR DESCRIPTION
**Pull request by driazati**
Stacked PRs
 * #35317 - [wip][jit] C++ API Implementation
 * **#34640 - [jit] C++ API Wrappers Proposal**

Currently we expose our internal APIs as the user APIs in C++. This leads to problems like internal assert failures when someone calls the wrong `.toX` on an `IValue`, even though the source of the assert failure is user error. There is also confusion about certain things like our [parameter/attribute iterators on module](https://discuss.pytorch.org/t/training-a-torch-module/58512/3) that would be good to clear up by providing a more idiomatic (and pybind-like) API.

You can browse the doxygen output here https://driazati.github.io/pytorch_doc_previews/34640/api/namespace_torch__jit__api.html#namespace-torch-jit-api

Looking for feedback on naming, functionality, consistency, etc.

NB: This doesn't include APIs for binding in list/dict, that will probably be a whole separate thing

### CODE REVIEW

File: torch/csrc/jit/api/api.h [hunk 1]
```diff
@@ -0,0 +1,358 @@
...
+using DictAccessor = Accessor<detail::DictAccessorPolicy>;
+
+struct List : public Value {
+  List(TypePtr type, std::initializer_list<Value> = {});
+  // list must not be empty
+  List(std::initializer_list<Value>);
+  ListAccessor operator[](size_t i) const;
+  size_t size();
+  iterable<Value> begin();
+  iterable<Value> end();
```
Comment (SplitInfinity) [inline torch/csrc/jit/api/api.h]: Did you mean to use `iterator` here?

**Review by SplitInfinity — COMMENTED** [review]
Consider adding methods to get `const iterators` for `List`, `Dict`, etc.

File: torch/csrc/jit/api/api.h [hunk 2]
```diff
@@ -0,0 +1,358 @@
...
+  typedef iterator self_type;
+  typedef T value_type;
+  typedef T& reference;
+  typedef T* pointer;
+  typedef std::forward_iterator_tag iterator_category;
+
+  iterator();
+
+  reference operator*() const;
+  pointer operator->() const;
```
[the remark this answers is not in the document]
Comment (driazati) [reply torch/csrc/jit/api/api.h]: For `operator->()` it's apparently different, it calls that then uses the result as the pointer to dereference normally. [This has a nice example](https://stackoverflow.com/questions/4928557/how-do-i-create-and-use-a-class-arrow-operator)

File: torch/csrc/jit/api/api.h [hunk 3]
```diff
@@ -0,0 +1,358 @@
...
+
+  self_type operator++();
+  self_type operator++(int);
```
[the remark this answers is not in the document]
Comment (driazati) [reply torch/csrc/jit/api/api.h]: It's just the postfix operator so we probably don't really need it

**Pull request closed without merging**